### PR TITLE
Fix unknownTypes handling

### PR DESCRIPTION
**Pull request by sj14**
Hi 👋  
my colleagues had the problem that wrong things were applied to the clusters. We figured out that qbec continues processing on certain errors (see log extract below):

```
...
[warn] error getting resources for type XYZ : an error on the server ("Internal Server Error") has prevented the request from succeeding
cluster metadata load took 1.816s
...
```

Digging into the code showed that the `resolver` struct contains a `err` variable at

https://github.com/splunk/qbec/blob/e0668af709b1211dfe2b3aafb53a7176340d5ffc/internal/remote/k8smeta/meta.go#L206

which is checked at

https://github.com/splunk/qbec/blob/e0668af709b1211dfe2b3aafb53a7176340d5ffc/internal/remote/k8smeta/meta.go#L290-L292

but, as far as I see, it was never assigned. Thus the error was only logged.

This PR assigns to the err variable which will hopefully solve the issue. Unfortunately, I was not yet able to reproduce the issue and I hope that we can already analyze or verify this issue while I try to reproduce it.

### CODE REVIEW

**Comment by gotwarlost** [comment]
Thanks for the bug report and PR.

Could you explain a little more about what happened when 

```
my colleagues had the problem that wrong things were applied to the clusters
```
what were the problems encountered?

This change is going to cause more problems than it solves. When developing qbec I tried to make it as fail-fast as possible but discovery is _such_ a rabbit hole that I wasn't able to do it for this case. 

* Due to the way discovery has evolved over time, errors happen every now and again and it is not always clear which errors are benign and which ones are not.

* Things like aggregated APIs that can contribute types can fail for the weirdest of reasons. In our cluster(s), we occasionally see internal server errors for pod metrics, for example.

* qbec wants to get every type in the system for the purposes of GC and it is not clear which types are important and which ones are not.

Because of the above, a qbec operation can be blocked because some type that is never used has errors. This _usually_ happens for types you don't necessarily care about.

I think we should have a conversation in #sig-api-machinery on the kubernetes slack on how to handle this. At any rate stopping the operation for a discovery problem that may or may not matter is problematic. It would be better if we tracked errors per type and only failed the operation is something needed a type that is known to have a problem.

**Comment by sj14** [comment]
> Could you explain a little more about what happened when

Sometimes, deployments get deleted when applying changes which were not shown while doing the diff. This seems to be related with a bad network connection to the cluster and Gravitational's Teleport in between (not really sure if this adds to the root cause).
Unfortunately, the logs I have are without a verbosity level. The given error was the only obvious issue to me. It might take me some time to get a deeper understanding of the topic and more detailed information.

**Comment by gotwarlost** [comment]
Ouch! Deployments getting deleted is really bad. Would love to get some additional context here. I've never seen something like this happen.

**Comment by sj14** [comment]
At least I can now reproduce it on our clusters. The problem seems to occur when this error is shown:

https://github.com/splunk/qbec/blob/13d7746613f84e91ec3f85e017e453673519655c/internal/commands/remote-list.go#L72

Creating a kind test cluster (with the qbec demo content) and modifying qbec to always return an error for this check, reproduces the issue.

I'm still not sure where the actual mistake in the code is and where to fix it.

**Comment by sj14** [comment]
Removing those lines seems to fix the issue for me with the combinations I tested, but I don't really understand all the consequences of the change yet and if this will introduce regressions.

https://github.com/splunk/qbec/blob/13d7746613f84e91ec3f85e017e453673519655c/internal/commands/remote-list.go#L132-L134

**Comment by sj14** [comment]
It seems like `unknownTypes` isn't used anywhere else.
@gotwarlost sorry to ping you but did you already have a change to look into the issue?

**Comment by gotwarlost** [comment]
sorry I haven't had the chance. Need to meet some deadlines at day job. Will take a look shortly

File: internal/commands/remote-list.go [hunk 1]
```diff
@@ -129,9 +127,6 @@ func (r *remoteLister) deletions(all []model.K8sLocalObject, filter func(obj mod
 
 	var removals []model.K8sQbecMeta
 	for _, c := range all {
-		if r.unknownTypes[c.GroupVersionKind()] {
```
Comment (harsimranmaan) [inline internal/commands/remote-list.go:132]: Maybe we can classify the unknownTypes rather than just eliminate them. If we can get hold of the API error(Internal Server Error, timeout etc.) while getting server objects, it might give us more mileage in solving the issue with unexpected deletions while also preserving the efficiency gains that we get by ignoring many errors. Maybe the client can use a new cli switch to toggle the behaviour while running qbec commands. A slightly different not-very-user-friendly way that I can think of is passing a list of api-resources in qbec.yaml that must be resolvable before we could do a certain deployment action in an env.

Comment (sj14) [reply internal/commands/remote-list.go:132]: Thanks a lot for your review. When I understand it correctly, you suggest classifying which kind of error happened at https://github.com/splunk/qbec/blob/13d7746613f84e91ec3f85e017e453673519655c/internal/commands/remote-list.go#L69 (Internal Server Error, timeout, etc) and based on this, it will `continue` here or not. Unfortunately, I haven't understood what's the purpose of the `unknownTypes` map yet. As far as I understand, it will always delete things on the remote site when there was an error at the referenced line, no matter which kind of classified error.  

In general, I agree with everything which solves the issue and I try to help as much as possible. Also feel free to modify this PR directly or create an own one when this might be easier/faster for you.

**Comment by gotwarlost** [comment]
OK, finally got around to looking at this problem. Sorry for the delay.

I had misunderstood the original problem.

The issue happens in the following case:

* Transient error causes discovery to not have a type used in the run and
* Remote listing of objects for this type still works
* The calculation of objects to delete does not remove local objects from the remote collection because it didn't find the type previously. This is absolutely wrong - the input list of objects that would be created by the source should be preserved even if the type is not found for it.

In other words, your fix is actually spot on and I will simply merge it.

**Comment by gotwarlost** [comment]
Actually, I now remember the reason why the unknownTypes check was there in the first place. It is to handle the case for custom resources when the CRD itself is not present (i.e. the case where a qbec app installs CRDs and custom resources for that CRD, or when an operator lazily creates the CRD). In this case, the removal of such an object from the remote collection will fail because of missing metadata causing qbec to halt prematurely instead of waiting for the crd to appear on the server

**Comment by sj14** [comment]
Thanks for merging the change. I don't have any expertise on CRDs right now, so I can't relate to this :-/
